web3.js/src/bpf-loader-upgradeable.ts

```diff
@@ -0,0 +1,699 @@
+        {pubkey: SYSVAR_CLOCK_PUBKEY, isSigner: false, isWritable: false},
+        {
+          pubkey: params.authorityPubkey,
+          isSigner: true,
+          isWritable: false,
+        },
+      ],
+      programId: this.programId,
+      data,
+    });
+  }
+
+  /**
+   * Generate a transaction instruction that set a new buffer authority
+   */
+  static setBufferAuthority(
+    params: SetBufferAuthorityParams,
+  ): TransactionInstruction {
+    const type = BPF_UPGRADEABLE_LOADER_INSTRUCTION_LAYOUTS.SetAuthority;
+    const data = encodeData(type, {});
+
+    return new TransactionInstruction({
+      keys: [
+        {pubkey: params.bufferPubkey, isSigner: false, isWritable: true},
+        {
+          pubkey: params.authorityPubkey,
+          isSigner: true,
+          isWritable: false,
+        },
+        {pubkey: params.newAuthorityPubkey, isSigner: false, isWritable: false},
+      ],
+      programId: this.programId,
+      data,
+    });
+  }
+
+  /**
+   * Generate a transaction instruction that set a new program authority
+   */
+  static async setUpgradeAuthority(
+    params: SetUpgradeAuthorityParams,
+  ): Promise<TransactionInstruction> {
+    const type = BPF_UPGRADEABLE_LOADER_INSTRUCTION_LAYOUTS.SetAuthority;
+    const data = encodeData(type, {});
+
+    const programDataPubkey = await this.getProgramDataAddress(
+      params.programPubkey,
+    );
+
+    const keys = [
+      {pubkey: programDataPubkey, isSigner: false, isWritable: true},
+      {
+        pubkey: params.authorityPubkey,
+        isSigner: true,
+        isWritable: false,
+      },
+    ];
+
+    if (params.newAuthorityPubkey) {
+      keys.push({
+        pubkey: params.newAuthorityPubkey,
+        isSigner: false,
+        isWritable: false,
+      });
+    }
+
+    return new TransactionInstruction({
+      keys,
+      programId: this.programId,
+      data,
+    });
+  }
+
+  /**
+   * Generate a transaction instruction that close program, buffer, or
+   *   uninitialized account
+   */
+  static close(params: CloseParams): TransactionInstruction {
+    const type = BPF_UPGRADEABLE_LOADER_INSTRUCTION_LAYOUTS.Close;
+    const data = encodeData(type, {});
+
+    const keys = [
+      {pubkey: params.closePubkey, isSigner: false, isWritable: true},
+      {
+        pubkey: params.recipientPubkey,
+        isSigner: false,
+        isWritable: true,
+      },
+    ];
+
+    if (params.authorityPubkey) {
+      keys.push({
+        pubkey: params.authorityPubkey,
+        isSigner: true,
+        isWritable: false,
+      });
+    }
+
+    if (params.programPubkey) {
+      keys.push({
+        pubkey: params.programPubkey,
+        isSigner: false,
+        isWritable: true,
+      });
+    }
+
+    return new TransactionInstruction({
+      keys,
+      programId: this.programId,
+      data,
+    });
+  }
+}
+
+/**
+ * BpfLoaderUpgradeable program interface
+ */
+export class BpfLoaderUpgradeable {
```

Comment:
@fanatid I would still like this to be removed

Reply:
Same as here, right? https://github.com/solana-labs/solana-web3.js/pull/1057#discussion_r756080092
Not sure why I re-opened PR here to be honest :laughing:

Reply:
Yeah exactly. Sorry I know it's not fun to remove code that you spent time on. It's just hard to maintain the upgrade process. Like I said before, a separate package for this would be welcome!